crates/store/src/state.rs

```diff
@@ -96,17 +96,107 @@ pub struct TransactionInputs {
+        self.0.add(el);
+    }
+
+    /// Returns an [`MmrProof`] for the leaf at the specified position
+    pub fn open(&self, pos: usize) -> Result<MmrProof, MmrError> {
+        self.0.open_at(pos, self.0.forest())
+    }
+
+    /// Returns the latest block number and partial mmr
+    pub fn chain_mmr(
+        &self,
+        blocks: &BTreeSet<BlockNumber>,
+        highest_block_number: &BlockNumber,
+    ) -> Result<(BlockNumber, PartialMmr), GetBatchInputsError> {
+        // Grab the block merkle paths from the inner state.
+        //
+        // NOTE: Scoped block to automatically drop the mutex guard asap.
+        //
+        // We also avoid accessing the db in the block as this would delay
+        // dropping the guard.
+        let (batch_reference_block, partial_mmr) = {
+            let latest_block_num = highest_block_number;
+
+            // Using latest block as the target forest means we take the state of the MMR one before
+            // the latest block. This is because the latest block will be used as the reference
+            // block of the batch and will be added to the MMR by the batch kernel.
+            let target_forest = latest_block_num.as_usize();
+            let peaks = self
+                .peaks_at(target_forest)
+                .expect("target_forest should be smaller than forest of the chain mmr");
+            let mut partial_mmr = PartialMmr::from_peaks(peaks);
+
+            for block_num in blocks.iter().map(BlockNumber::as_usize) {
+                // SAFETY: We have ensured block nums are less than chain length.
+                let leaf = self
+                    .0
+                    .get(block_num)
+                    .expect("block num less than chain length should exist in chain mmr");
+                let path = self
+                    .0
+                    .open_at(block_num, target_forest)
+                    .expect("block num and target forest should be valid for this mmr")
+                    .merkle_path;
+                // SAFETY: We should be able to fill the partial MMR with data from the chain MMR
+                // without errors, otherwise it indicates the chain mmr is invalid.
+                partial_mmr
+                    .track(block_num, leaf, &path)
+                    .expect("filling partial mmr with data from mmr should succeed");
+            }
+
+            (latest_block_num, partial_mmr)
+        };
+        Ok((*batch_reference_block, partial_mmr))
+    }
+}
 /// Container for state that needs to be updated atomically.
 struct InnerState {
     nullifier_tree: NullifierTree,
-    chain_mmr: Mmr,
+    blockchain: Blockchain,
     account_tree: SimpleSmt<ACCOUNT_TREE_DEPTH>,
 }
 
 impl InnerState {
     /// Returns the latest block number.
     fn latest_block_num(&self) -> BlockNumber {
-        let block_number: u32 = (self.chain_mmr.forest() - 1)
+        let block_number: u32 = (self.blockchain.0.forest() - 1)
```

Comment:
The idea of adding methods like `Blockchain::chain_tip` is that we can call them here, so we can avoid accessing the inner `Mmr` directly. So all `blockchain.0.xyz()` methods should be replaced by calling methods directly, e.g. `blockchain.xyz()` (or with blockchain-specific names, so `blockchain.chain_tip()` instead of `blockchain.0.forest()`.

Here we should be able to call `chain_tip`, right?